vector_setters.go

```diff
@@ -32,108 +33,319 @@ func setPrimitive[T any](vec *vector, rowIdx C.idx_t, v T) {
+	}
+	setPrimitive(vec, rowIdx, fv)
+	return nil
+}
+func setTS[S any](vec *vector, t Type, rowIdx C.idx_t, val S) error {
+	var ti time.Time
+	switch v := any(val).(type) {
+	case time.Time:
+		ti = v
+	default:
+		return castError(reflect.TypeOf(val).String(), reflect.TypeOf(t).String())
+	}
 	var ticks int64
 	switch t {
 	case TYPE_TIMESTAMP:
-		ticks = v.UTC().UnixMicro()
+		ticks = ti.UTC().UnixMicro()
 	case TYPE_TIMESTAMP_S:
-		ticks = v.UTC().Unix()
+		ticks = ti.UTC().Unix()
 	case TYPE_TIMESTAMP_MS:
-		ticks = v.UTC().UnixMilli()
+		ticks = ti.UTC().UnixMilli()
 	case TYPE_TIMESTAMP_NS:
-		ticks = v.UTC().UnixNano()
+		ticks = ti.UTC().UnixNano()
 	case TYPE_TIMESTAMP_TZ:
-		ticks = v.UTC().UnixMicro()
+		ticks = ti.UTC().UnixMicro()
 	}
-
 	var ts C.duckdb_timestamp
 	ts.micros = C.int64_t(ticks)
 	setPrimitive(vec, rowIdx, ts)
+	return nil
 }
-
-func (vec *vector) setDate(rowIdx C.idx_t, val any) {
-	// Days since 1970-01-01.
-	v := val.(time.Time)
-	days := int32(v.UTC().Unix() / secondsPerDay)
-
-	var date C.duckdb_date
-	date.days = C.int32_t(days)
-	setPrimitive(vec, rowIdx, date)
+func setDate[S any](vec *vector, rowIdx C.idx_t, val S) error {
+	var date time.Time
+	switch v := any(val).(type) {
+	case time.Time:
+		date = v
+	default:
+		return castError(reflect.TypeOf(val).String(), reflect.TypeOf(date).String())
+	}
+	days := int32(date.UTC().Unix() / secondsPerDay)
+	var date2 C.duckdb_date
+	date2.days = C.int32_t(days)
+	setPrimitive(vec, rowIdx, date2)
+	return nil
 }
 
-func (vec *vector) setTime(rowIdx C.idx_t, val any) {
-	v := val.(time.Time)
-	ticks := v.UTC().UnixMicro()
+func setTime[S any](vec *vector, rowIdx C.idx_t, val S) error {
+	var date time.Time
+	switch v := any(val).(type) {
+	case time.Time:
+		date = v
+	default:
+		return castError(reflect.TypeOf(val).String(), reflect.TypeOf(date).String())
+	}
+	ticks := date.UTC().UnixMicro()
 
 	var t C.duckdb_time
 	t.micros = C.int64_t(ticks)
 	setPrimitive(vec, rowIdx, t)
+	return nil
 }
 
-func (vec *vector) setInterval(rowIdx C.idx_t, val any) {
-	v := val.(Interval)
-	var interval C.duckdb_interval
-	interval.days = C.int32_t(v.Days)
-	interval.months = C.int32_t(v.Months)
-	interval.micros = C.int64_t(v.Micros)
-	setPrimitive(vec, rowIdx, interval)
+func setInterval[S any](vec *vector, rowIdx C.idx_t, val S) error {
+	var interval Interval
+	switch v := any(val).(type) {
+	case Interval:
+		interval = v
+	default:
+		return castError(reflect.TypeOf(val).String(), reflect.TypeOf(interval).String())
+	}
+	var interval2 C.duckdb_interval
+	interval2.days = C.int32_t(interval.Days)
+	interval2.months = C.int32_t(interval.Months)
+	interval2.micros = C.int64_t(interval.Micros)
+	setPrimitive(vec, rowIdx, interval2)
+	return nil
 }
 
-func (vec *vector) setHugeint(rowIdx C.idx_t, val any) {
-	v := val.(*big.Int)
-	hugeInt, _ := hugeIntFromNative(v)
-	setPrimitive(vec, rowIdx, hugeInt)
+func setHugeint[S any](vec *vector, rowIdx C.idx_t, val S) error {
```

Comment:
Also here, why are we adding the extra cases?
And, since we return errors now, we do not need to omit the error in `fv, _ = hugeIntFromNative(v)` any longer.
```Go
func setHugeint[S any](vec *vector, rowIdx C.idx_t, val S) error {
	var fv C.duckdb_hugeint
	switch v := any(val).(type) {
	case *big.Int:
		if v == nil {
			return castError(reflect.TypeOf(val).String(), reflect.TypeOf(fv).String())
		}
		fv, _ = hugeIntFromNative(v)
	default:
		return castError(reflect.TypeOf(val).String(), reflect.TypeOf(fv).String())
	}
	setPrimitive(vec, rowIdx, fv)
	return nil
}
```

Reply:
same as above

Reply:
Actually I think it would be nice to be able to set hugeints with normal ints, working with `big.Int` can be cumbersome and slow, if the user knows they won't hit the limit of normal ints it would be nice if we don't force them to use `big.Int`. But I will add test cases for this

Reply:
Yes, I agree. Cool.